core/src/Kokkos_Parallel.hpp

```diff
@@ -112,16 +90,14 @@ using detected_device_type_t = typename is_detected_device_type<T>::type;
 
 template <class Functor, class Policy>
 struct FunctorPolicyExecutionSpace {
-  using execution_space = std::conditional_t<
-      is_detected_execution_space<Policy>::value,
-      detected_execution_space_t<Policy>,
-      std::conditional_t<
-          is_detected_execution_space<Functor>::value,
-          detected_execution_space_t<Functor>,
+  using execution_space = detected_or_t<
+      detected_or_t<
           std::conditional_t<
-              is_detected_device_type<Functor>::value,
-              detected_execution_space_t<detected_device_type_t<Functor>>,
-              Kokkos::DefaultExecutionSpace>>>;
+              is_detected<device_type_t, Functor>::value,
+              detected_t<execution_space_t, detected_t<device_type_t, Functor>>,
+              Kokkos::DefaultExecutionSpace>,
```

Comment:
Could rewrite this as
```
detected_or_t<Kokkos::DefaultExecutionSpace, device_type_t, Functor>::execution_space,
```

Reply:
You mean taking advantage that execution space define `<SomeExecSpace>::execution_space = <SomeExecSpace>`?

Reply:
Yes, in particular, `Kokkos::DefaultExecutionSpace::execution-space = Kokkos::DefaultExecutionSpace`.

Reply:
You want me to change or just pointing that out?

Reply:
I like my suggestion better since it's shorter and consistent with the style around it.

Reply:
This yields incomplete type error on a couple builds.
https://cloud.cees.ornl.gov/jenkins-ci/blue/organizations/jenkins/Kokkos/detail/Kokkos/4939/pipeline/40

I will revert 2a650c8.